### PR TITLE
Add a script to retrieve images by digest and put in install_bootstrap

### PR DESCRIPTION
**Pull request by david4096**
After some back and forth I am moving this script to where I believe it ought to belong so that it verifies our images on deploy. This PR will require going forward that all builds have uploaded their image digest to S3 via CircleCI.

* Gather the commit information for a tag or branch from GitHub
* Download the digest for the image made in CircleCI (from S3) https://github.com/dockstore/dockstore/pull/4274
* Add the script to install_bootstrap to gather the 
* ~~Pull the docker image and inspect the digest~~
* ~~Compare the two digests and exit with status code 1 for errors or failures~~
* ~~Add this script into the final stages of `install_bootstrap` before the deploy is done.~~

Having the ability to verify using branches should make it easier to test this in dev, since the dev redeploy happens against develop branch.

This is NOT intended for the in-progress release.

For previous conversation see https://github.com/dockstore/dockstore-deploy/pull/257

### CODE REVIEW

**Review by coverbeck — COMMENTED** [review]
What if instead we just enforce that the tag for the web service has a digest, with an exception for develop?

1. I think we need the exception for develop for both this PR's approach and the proposal to enforce a digest. There's the chance, almost but not entirely impossible given how small the window is, that between fetching the commit, the Docker image, and the digest, that they change in the middle of things.
2. The pros of ensuring we have a digest are that it's easier to check, and that it's following the best practice we are recommending to workflow developers.
3. The con is that we would have to hunt down the digest and we might accidentally select the wrong one, although your published digest should make that easier.

Not 100% sold on my proposal; throwing it out for consideration.

File: scripts/verify-webservice-image-digest.py
Comment (Richard-Hansen) [inline scripts/verify-webservice-image-digest.py:38]: Suggesting a short comment to explain the indexing, which I believe is from:
1. len('sha256:') == 7
2. len('sha256:') + (256 bits * (1 hex character / 4 bits)) = 71

File: scripts/verify-webservice-image-digest.py
Comment (Richard-Hansen) [inline scripts/verify-webservice-image-digest.py:40]: Very small (ignore-ably small) nit: camelCase amongst the snake case

**Comment by david4096** [comment]
> What if instead we just enforce that the tag for the web service has a digest, with an exception for develop?

That's fine with me, I think instead of this verification script we can pull the digest from the S3 location to automate away that step. It's not 100% clear to me what the other ramifications are yet so I moved this to a draft to consider.

File: scripts/verify-webservice-image-digest.py
Comment (david4096) [reply scripts/verify-webservice-image-digest.py]: It's a good suggestions but this part got removed since the verification is implicit now.

File: scripts/verify-webservice-image-digest.py
Comment (david4096) [reply scripts/verify-webservice-image-digest.py]: Thanks

**Comment by david4096** [comment]
I would like some help testing this but with the release it is low priority. I can't seem to run the develop branch of compose_setup (or this one) locally using `install_bootstrap` getting an error when getting the public IP address.

```curl: (7) Failed to connect to 169.254.169.254 port 80: No route to host
Shooot, could not parse view as JSON.
Tips: functions are not valid JSON and keys / values must be surround with double quotes.

SyntaxError: Unexpected token = in JSON at position 392
    at JSON.parse (<anonymous>)
    at parseView (/usr/local/lib/node_modules/mustache/bin/mustache:74:17)
    at onDone (/usr/local/lib/node_modules/mustache/bin/mustache:67:10)
    at ReadStream.onEnd (/usr/local/lib/node_modules/mustache/bin/mustache:118:5)
    at Object.onceWrapper (events.js:482:28)
    at ReadStream.emit (events.js:388:22)
    at endReadableNT (internal/streams/readable.js:1336:12)
    at processTicksAndRejections (internal/process/task_queues.js:82:21)
```

**Comment by denis-yuen** [comment]
> I can't seem to run the develop branch of compose_setup (or this one) locally using `install_bootstrap` getting an error when getting the public IP address.

I don't think this this will work locally, it gets the ip address by ec2 instance metadata 
 https://github.com/dockstore/compose_setup/blob/develop/install_bootstrap#L417

**Comment by coverbeck** [comment]
> > I can't seem to run the develop branch of compose_setup (or this one) locally using `install_bootstrap` getting an error when getting the public IP address.
> 
> I don't think this this will work locally, it gets the ip address by ec2 instance metadata
> https://github.com/dockstore/compose_setup/blob/develop/install_bootstrap#L417
> 
> (i.e. you might just want to run this on a toy VM in dev)

You could also create a Dockstore stack via the deployer. Let me know if you want to do that and I can walk you through it.

**Comment by david4096** [comment]
I ended up testing this on an instance and running `install_bootstrap` manually (not using the deployer) and found some unrelated issues that are addressed in the last two commits.

**Comment by david4096** [comment]
Putting this back as a draft, I overlooked how -deploy would need to interact with this.

**Comment by david4096** [comment]
Putting this back in review after some wrestling with GitHub actions and getting it working with how -deploy specifies a branch or tag.

**Review by coverbeck — COMMENTED** [review]
This looks really good; we can just specify a tag/branch and it will autofetch the correct digest. 

Did you test that the digest gets overwritten when subsequent builds of a branch are made? Should have asked that in your other PR, but that will need to work for our develop branch deploys.

**Review by denis-yuen — APPROVED** [review]
With @coverbeck comments above

File: install_bootstrap
Comment (david4096) [reply install_bootstrap:439]: > Did you test that the digest gets overwritten when subsequent builds of a branch are made? Should have asked that in your other PR, but that will need to work for our develop branch deploys.

It will pull the latest commit on a branch and check if its been uploaded to S3.

Comment (david4096) [reply install_bootstrap:439]: > Whoops, how did I do that and I'm surprised it didn't break things.

None of the tests check the interactive part of install_bootstrap, everything uses --script with a premade mustache json.

File: install_bootstrap
Comment (garyluu) [inline install_bootstrap:208]: two wget installs?

File: install_bootstrap
Comment (coverbeck) [reply install_bootstrap:435]: > > Whoops, how did I do that and I'm surprised it didn't break things.
> 
> None of the tests check the interactive part of install_bootstrap, everything uses --script with a premade mustache json.

I meant I'm surprised the CF template that relies on this didn't fail, but now that I think about it, the deployer writes out its own JSON

Comment (coverbeck) [reply install_bootstrap:435]: > > Did you test that the digest gets overwritten when subsequent builds of a branch are made? Should have asked that in your other PR, but that will need to work for our develop branch deploys.
> 
> It will pull the latest commit on a branch and check if its been uploaded to S3.

Ah, I see, you upload/download to `<branch>-<commit>`; so builds' digests won't overwrite each other.

File: install_bootstrap
Comment (coverbeck) [inline install_bootstrap:198]: Formatting is off, but if you just remove this (Gary's comment about installing wget twice)

**Review by coverbeck — APPROVED** [review]

**Review by garyluu — APPROVED** [review]

**Comment by david4096** [comment]
Tests are failing because a merge to dockstore@develop happened and we are still waiting for an s3 upload. Will merge after the most recent merge builds successfully.

**Pull request merged**
